File: Java/src/com/dbc/algorithm/String/LongestCommonSubsequence.java
package com.dbc.algorithm.String;

public class LongestCommonSubsequence {
    public int longestCommonSubsequence(String text1, String text2) {
        int[] dp = new int[text2.length() + 1];
        int result = 0;
        for (int i = 0; i < text1.length(); i++) {
            int front = 0;
            for (int j = 1; j < text2.length() + 1; j++) {
                int temp = dp[j];
                if (text1.charAt(i) == text2.charAt(j - 1)) {
                    dp[j] = front + 1;
                } else {
                    dp[j] = Math.max(dp[j - 1], Math.max(front, dp[j]));
                }
                front = temp;
                result = Math.max(result, dp[j]);
            }
        }
        return result;
    }
}

File: src/tmp/Python/string/longest_common_subsequence.py
def longestCommonSubsequence(text1: str, text2: str) -> int:
    m, n = len(text1), len(text2)
    dp = [0 for _ in range(n + 1)]
    res = 0
    for i in range(m):
        front = 0
        for j in range(1, n + 1):
            temp = dp[j]
            if text1[i] == text2[j - 1]:
                dp[j] = front + 1
            else:
                dp[j] = max(dp[j - 1], front, dp[j])
            front = temp
            res = max(res, dp[j])
    
    return res

if __name__ == "__main__":
    print(longestCommonSubsequence(text1 = "abc", text2 = "def"))
                